### ISSUE TITLE
Make custom pydantic color validator

### ISSUE THREAD

**Issue by sofroniewn** [napari/napari#3884]
## 🧰 Task

Following discussion in https://github.com/napari/napari/pull/3879#issuecomment-1000922569 we should make a custom color [pydantic type with its own validator](https://pydantic-docs.helpmanual.io/usage/types/#classes-with-__get_validators__) based on

https://github.com/napari/napari/blob/43aa65bcde1ef36493ba51499edca4fad8d7fafe/napari/components/text_overlay.py#L37-L39

and then reuse it in all places where we currently have

https://github.com/napari/napari/blob/43aa65bcde1ef36493ba51499edca4fad8d7fafe/napari/components/text_overlay.py#L32

or https://github.com/napari/napari/blob/9152dbd7c328b12a9c3b19c9629f58906650465c/napari/layers/surface/wireframe.py#L24

This will mean we no longer need to use the union of types, but always have one nice type

**Comment by kevinyamauchi** [comment]
Great idea!

**Comment by andy-sweet** [comment]
I was planning to do something like this in the style encoding work.

```python
class ColorValue(np.ndarray):
    """A 4x1 array that represents one RGBA color value."""

    @classmethod
    def __get_validators__(cls):
        yield cls.validate_type

    @classmethod
    def validate_type(cls, val):
        return transform_color(val)[0]


class ColorArray(np.ndarray):
    """An Nx4 array where each row of N represents one RGBA color value."""

    @classmethod
    def __get_validators__(cls):
        yield cls.validate_type

    @classmethod
    def validate_type(cls, val):
        return np.empty((0, 4)) if len(val) == 0 else transform_color(val)
```

See [the closed WIP PR](https://github.com/napari/napari/pull/3493/files#diff-c6c515811f47fb2a55c3edb7f778cc93b0bbf1435d21c11a11e78a3d6d1e7ba5R23) for more context.

At the moment we have a need to distinguish between a field that stores a single color and one that stores multiple colors, thus the two types.

One downside of this approach is that `Array[float, (4,)]` is more descriptive than `ColorValue` - i.e. you need to go look at the docstring and definition of `ColorValue` to understand it.

**Comment by sofroniewn** [comment]
I like `ColorValue` and `ColorArray`, mirrors what vispy has a bit of I think. What do @tlambert03 and @jni think?

**Comment by tlambert03** [comment]
yeah, I like `ColorValue` and `ColorArray`.  Perhaps a slight preference for just `Color` over `ColorValue` (then just use `pydantic.types.Color` when we need to differentiate between theirs and ours).  

> One downside of this approach is that Array[float, (4,)] is more descriptive than ColorValue

not too worried about this.

@andy-sweet ... can you remind me whether we can't just subclass pydantic's color to deal with the edge cases we have in napari?  I know that pydantic's color wasn't sufficient, but can't remember whether a subclass would fix it.

**Comment by andy-sweet** [comment]
> @andy-sweet ... can you remind me whether we can't just subclass pydantic's color to deal with the edge cases we have in napari? I know that pydantic's color wasn't sufficient, but can't remember whether a subclass would fix it.

I wrote [some notes about pydantic's `Color` limitations with respect to napari](https://hackmd.io/K5rCILaHQmWsNhY0ml_xaQ#Pydantic%E2%80%99s-Color). If we only wanted to represent single colors (and not multiple colors in one field), I think we could do what you said pretty easily.

The harder problem is how to handle fields that store multiple colors and how to get an Nx4 RGBA array (e.g. as vispy wants) efficiently. Using something like `List[Color]` directly probably means we'd have to iterate over that and copy it into an Nx4 array every time we read the field. I suppose we could define a `ColorArray` type and cache the Nx4 RGBA array.

I did try [using pydantic's color to reimplement `transform_color` in a branch](https://github.com/napari/napari/compare/main...andy-sweet:pydantic-color), but it wasn't as straightforward as I hoped. Maybe it's worth picking that work back up again as part of this issue.

Overall, I like the simplicity of the RGBA array representation. So given the extra work needed to make pydantic work, my recommendation is to stick with the RGBA array.
